### PR TITLE
fix: refactor `sync_headers_in_reverse` to use a standalone function for handling network tipsets

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

To address

              Could we avoid the inner/outer loop complexity by refactoring this to smaller functions that could potentially be unit tested? At least the block `for tipset in network_tipsets {`

_Originally posted by @LesnyRumcajs in https://github.com/ChainSafe/forest/pull/4459#discussion_r1660713502_
            

Changes introduced in this pull request:

- use a standalone function `for_each_tipset_until_epoch_overreached` for handling network tipsets
- more comments explaining the loop breaking logic

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/chain_sync/tipset_syncer.rs [hunk 1]
```diff
@@ -851,15 +851,20 @@ async fn sync_headers_in_reverse<DB: Blockstore + Sync + Send + 'static>(
...
             validate_tipset_against_cache(bad_block_cache, tipset.key(), &accepted_blocks)?;
             accepted_blocks.extend(tipset.cids());
             tracker.write().set_epoch(tipset.epoch());
             pending_tipsets.push(tipset);
+            Ok(())
+        };
+        // Breaks the loop when `until_epoch` is overreached, which happens
+        // when there are null tipsets in the queried range.
+        // Note that when the `until_epoch` is null, the outer while condition
+        // is always true, and it relies on this flag to break the loop
```
Comment (LesnyRumcajs) [inline src/chain_sync/tipset_syncer.rs:864]: Does this comment match the implemented logic?

Comment (hanabi1224) [reply src/chain_sync/tipset_syncer.rs:864]: Yes. Which part do you think is in doubt?

Comment (LesnyRumcajs) [reply src/chain_sync/tipset_syncer.rs:864]: What is the `flag`?

Comment (hanabi1224) [reply src/chain_sync/tipset_syncer.rs:864]: Oh, it refers to the returned boolean value. How about 
`and it relies on the returned boolean flag(until epoch is overreached) to break the loop`

Comment (hanabi1224) [reply src/chain_sync/tipset_syncer.rs]: fixed.

Comment (LesnyRumcajs) [reply src/chain_sync/tipset_syncer.rs]: I wouldn't use the word `flag`; it suggests (subjectively) that there is a mutable, boolean variable flying around upon which the outer loop relies. `return value` would be more fitting here.

**Review by LesnyRumcajs — APPROVED** [review]

File: src/chain_sync/tipset_syncer.rs (hunk 1 above)
Comment (hanabi1224) [reply src/chain_sync/tipset_syncer.rs]: Fixed

**Review by ruseinov — APPROVED** [review]

**Pull request merged**